### PR TITLE
fix: Add Tick to Show Me > Electron APIs

### PR DESCRIPTION
**Pull request by JordanTroutman**
Addressing Issue #282.  I added functionality to put a checkmark that keeps track of the current open Fiddle in `Show Me > Electron APIs`.  Each item within the `MenuItem` has the `radio` type.

### CODE REVIEW

**Review by erickzhao — CHANGES_REQUESTED** [review]
Hey @JordanTroutman!

So this is a good start, but the `radio` MenuItem type unfortunately doesn't fulfill all of our requirements here. A few acceptance criteria that need to be done:

1) There should be no active checkbox when no Demo is selected.

Unfortunately, this is a limitation in the Chromium model for the `radio` model (see https://github.com/electron/electron/issues/21292#issuecomment-561923188), so this means that the checkboxes need to be managed programatically.

2) When loading another Fiddle, the checkboxes should be reset.

If we load a Fiddle from another source (remote Gist, New Fiddle, or local folder), the ticked MenuItem should be cleared.

3) When loading more than 1 window, the checkboxes should be adjusted accordingly.

Let's say we have two Electron Fiddle windows opened, one of them on the Debugger demo and one on the TouchBar demo, swapping between windows should tick the correct MenuItem for each window.

File: src/main/menu.ts [hunk 1]
```diff
@@ -150,9 +150,12 @@ function getTasksMenu(): MenuItemConstructorOptions {
 function getShowMeMenuItem(
   key: string,
   item: string | Templates,
+  active_key: string | null,
```
Comment (vhashimotoo) [inline src/main/menu.ts:153]: Let's use camelCase

```suggestion
  activeKey: string | null,
```

File: src/main/menu.ts [hunk 2]
```diff
@@ -150,9 +150,12 @@ function getTasksMenu(): MenuItemConstructorOptions {
...
 ): MenuItemConstructorOptions {
   if (typeof item === 'string') {
     return {
+      type: 'checkbox',
+      checked: key == active_key,
```
Comment (vhashimotoo) [inline src/main/menu.ts:158]: Why we can't use `===` here?

```suggestion
      checked: key == activeKey,
```

File: src/main/menu.ts [hunk 3]
```diff
@@ -161,15 +164,15 @@ function getShowMeMenuItem(
   return {
     label: key,
     submenu: Object.keys(item).map((subkey) => {
-      return getShowMeMenuItem(subkey, item[subkey]);
+      return getShowMeMenuItem(subkey, item[subkey], active_key);
```
Comment (vhashimotoo) [inline src/main/menu.ts:167]:
```suggestion
      return getShowMeMenuItem(subkey, item[subkey], activeKey);
```

File: src/main/menu.ts [hunk 4]
```diff
@@ -161,15 +164,15 @@ function getShowMeMenuItem(
...
     }),
   };
 }
 
-function getShowMeMenu(): MenuItemConstructorOptions {
+function getShowMeMenu(active_key: string | null): MenuItemConstructorOptions {
```
Comment (vhashimotoo) [inline src/main/menu.ts:172]:
```suggestion
function getShowMeMenu(activeKey: string | null): MenuItemConstructorOptions {
```

File: src/main/menu.ts [hunk 5]
```diff
@@ -161,15 +164,15 @@ function getShowMeMenuItem(
...
   const showMeMenu: Array<MenuItemConstructorOptions> = Object.keys(
     SHOW_ME_TEMPLATES,
-  ).map((key) => getShowMeMenuItem(key, SHOW_ME_TEMPLATES[key]));
+  ).map((key) => getShowMeMenuItem(key, SHOW_ME_TEMPLATES[key], active_key));
```
Comment (vhashimotoo) [inline src/main/menu.ts:175]:
```suggestion
  ).map((key) => getShowMeMenuItem(key, SHOW_ME_TEMPLATES[key], activeKey));
```

File: src/main/menu.ts [hunk 6]
```diff
@@ -248,7 +251,7 @@ function getFileMenu(): MenuItemConstructorOptions {
 /**
  * Creates the app's window menu.
  */
-export function setupMenu() {
+export function setupMenu(active_key: string | null = null) {
```
Comment (vhashimotoo) [inline src/main/menu.ts:254]:
```suggestion
export function setupMenu(activeKey: string | null = null) {
```

File: src/main/menu.ts [hunk 7]
```diff
@@ -331,7 +334,7 @@ export function setupMenu() {
 
   menu.splice(process.platform === 'darwin' ? 1 : 0, 0, getFileMenu());
 
-  menu.splice(menu.length - 1, 0, getTasksMenu(), getShowMeMenu());
+  menu.splice(menu.length - 1, 0, getTasksMenu(), getShowMeMenu(active_key));
```
Comment (vhashimotoo) [inline src/main/menu.ts:337]:
```suggestion
  menu.splice(menu.length - 1, 0, getTasksMenu(), getShowMeMenu(activeKey));
```

File: src/main/menu.ts (hunk 6 above)
Comment (vhashimotoo) [reply src/main/menu.ts:254]: Also since `activeKey` is unused under `index.ts`, why not use `activeKey?: string | null`

**Review by vhashimotoo — COMMENTED** [review]
Since I'm not really known the history of the request so I'm just reporting some additions and bug report:

When you open two windows and choose on the first some another example, and on the second another example that not on window 1, you get that both windows have selected examples from 1 or 2.

**Comment by JordanTroutman** [comment]
@vhashimotoo Can you explain a little more of exactly what you tried? I just opened two windows and clicked a different example for each window.  Focusing on a window ticked the appropriate example.

**Comment by vhashimotoo** [comment]
This seems reproduced very weirdly, you can check the gif below. You need to have two windows and on the first window select example, second shouldn't take the value of the first, but if you select example on the second window the value is changed also in the first window.

recording

**Comment by JordanTroutman** [comment]
electron-gif
This is what it looked like on my Macbook...is there a difference between OS when using menus?

**Comment by erickzhao** [comment]
@JordanTroutman I can also repro on my Windows VM. I'll try to investigate a bit deeper soon.

File: src/main/windows.ts [hunk 8]
```diff
@@ -54,6 +54,19 @@ export function createMainWindow(): Electron.BrowserWindow {
     browserWindow = null;
   });
 
+  // Focusing on new window loads the possibly clicked "Show Me" fiddle
+  browserWindow.on('focus', () => {
+    browserWindow?.webContents.send(IpcEvents.MENU_IPC_RENDER);
```
Comment (erickzhao) [inline src/main/windows.ts:59]: Can we use the `IpcMainManager` and `IpcRendererManager` modules to handle IPC events?

File: src/main/windows.ts [hunk 9]
```diff
@@ -54,6 +54,19 @@ export function createMainWindow(): Electron.BrowserWindow {
...
+  });
+
+  // Show menu for respective window taking into consideration an entry in
+  // "Show Me" may be selected
+  ipcMain.on(IpcEvents.MENU_IPC_MAIN, (_, result) => {
```
Comment (erickzhao) [inline src/main/windows.ts:64]: This IPC event will duplicate on each window creation. Should probably go elsewhere.

**Review by erickzhao — CHANGES_REQUESTED** [review]
A few general comments, unrelated to the Windows bug pointed above.

**Comment by erickzhao** [comment]
superseded by #705

**Pull request closed without merging**